conformance/runner.js

```diff
@@ -0,0 +1,321 @@
+const document = require('../src/document')(reference.DocumentReference);
+const DocumentSnapshot = document.DocumentSnapshot;
+const convert = require('../src/convert');
+const ResourcePath = require('../src/path').ResourcePath;
+
+/** Loads the Protobuf definition and the binary test data. */
+function loadTestCases() {
+  const protobufRoot = new protobufjs.Root();
+
+  protobufRoot.resolvePath = function(origin, target) {
+    if (/^google\/.*/.test(target)) {
+      target = path.join(googleProtoFiles(), target.substr('google/'.length));
+    }
+    return target;
+  };
+
+  const protoDefinition = protobufRoot.loadSync(
+    path.join(__dirname, 'test_definition.proto')
+  );
+
+  const binaryProtoData = require('fs').readFileSync(
+    path.join(__dirname, 'test_data.binprotos')
+  );
+
+  const testType = protoDefinition.lookupType('tests.Test');
+
+  const testCases = [];
+
+  for (let offset = 0; offset < binaryProtoData.length; ) {
+    const messageLength = varint.decode(binaryProtoData, offset);
+    offset += varint.encodingLength(messageLength);
+
+    let testCase = testType.decode(
+      binaryProtoData.slice(offset, offset + messageLength)
+    );
+    testCases.push(testCase);
+
+    offset = offset + messageLength;
+  }
+
+  return testCases;
+}
+
+/** Converst a JSON object into a JS Object suitable for the Node API. */
+function convertInput(json) {
+  const obj = JSON.parse(json);
+
+  function convertValue(value) {
+    if (is.object(value)) {
+      return convertObject(value);
+    } else if (is.array(value)) {
+      return convertArray(value);
+    } else if (value === 'Delete') {
+      return Firestore.FieldValue.delete();
+    } else if (value === 'ServerTimestamp') {
+      return Firestore.FieldValue.serverTimestamp();
+    }
+
+    return value;
+  }
+
+  function convertArray(arr) {
+    for (let i = 0; i < arr.length; ++i) {
+      arr[i] = convertValue(arr[i]);
+    }
+    return arr;
+  }
+
+  function convertObject(obj) {
+    for (const key in obj) {
+      if (obj.hasOwnProperty(key)) {
+        obj[key] = convertValue(obj[key]);
+      }
+    }
+    return obj;
+  }
+
+  return convertValue(obj);
+}
+
+/** Converts a CommitRequest in Proto3 JSON to Protobuf JS format. */
+function convertCommit(commitRequest) {
+  const deepCopy = JSON.parse(JSON.stringify(commitRequest));
+  for (let write of deepCopy.writes) {
+    if (write.update) {
+      write.update.fields = convert.documentFromJson(write.update.fields);
+    }
+  }
+
+  return deepCopy;
+}
+
+/** Converts a Protobuf Precondition to the expected user input. */
+function convertPrecondition(precondition) {
+  const result = JSON.parse(JSON.stringify(precondition));
+  if (result.updateTime) {
+    result.lastUpdateTime = DocumentSnapshot.toISOTime(result.updateTime);
+    delete result.updateTime;
+  }
+  return result;
+}
+
+/** List of test cases that are ignored. */
+const ignored = [
+  // Firestore Node does not omit empty writes
```

Comment:
I was thinking that it would be simpler in the long run if all the clients generated the same, optimal requests.

Reply:
Yes, I think that's a good optimization. It might get a little tricky to only return one WriteResult with potentially one or two requests per write, but I do agree that it is the right way forward.

Reply:
I just always return the first WriteResult. Is that bad if there is only a transform?

Reply:
I'm not doing this correctly right now in Node, but I think we should return the result from the write transform (whether there is one or two). If an update doesn't change the document, but the transform does, then the WriteResult for the update will be stale and the updateTime will only be correctly reflected in the document transform.

Are you returning WriteResults from WriteBatches? It gets a little trickier to filter out these extra WriteResults in WriteBatches.